File: flutter/test-app-local-dependency/lib/other_tab.dart
/*
 * Copyright (c) 2018-2022 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import 'dart:async';
import 'dart:io';

import 'package:ffmpeg_kit_flutter/ffmpeg_kit.dart';
import 'package:ffmpeg_kit_flutter/ffmpeg_kit_config.dart';
import 'package:ffmpeg_kit_flutter/log.dart';
import 'package:ffmpeg_kit_flutter/return_code.dart';
import 'package:flutter/material.dart';

import 'abstract.dart';
import 'popup.dart';
import 'tooltip.dart';
import 'util.dart';
import 'video_util.dart';

class OtherTab {
  static const String DAV1D_TEST_DEFAULT_URL =
      "http://download.opencontent.netflix.com.s3.amazonaws.com/AV1/Sparks/Sparks-5994fps-AV1-10bit-960x540-film-grain-synthesis-854kbps.obu";

  late Refreshable _refreshable;
  late String _selectedTest;
  String _outputText = "";

  void init(Refreshable refreshable) {
    _refreshable = refreshable;
    List<DropdownMenuItem<String>> testList = getTestList();
    _selectedTest = testList[0].value!;
    this.clearOutput();
  }

  void changedTest(String? selectedTest) {
    _selectedTest = selectedTest!;
    _refreshable.refresh();
  }

  void setActive() {
    print("Other Tab Activated");
    FFmpegKitConfig.enableLogCallback(null);
    FFmpegKitConfig.enableStatisticsCallback(null);
    showPopup(OTHER_TEST_TOOLTIP_TEXT);
  }

  void logCallback(Log log) {
    this.appendOutput(log.getMessage());
  }

  void appendOutput(String logMessage) {
    _outputText += logMessage;
    _refreshable.refresh();
  }

  void clearOutput() {
    _outputText = "";
    _refreshable.refresh();
  }

  void runTest() {
    this.clearOutput();

    switch (this._selectedTest) {
      case "chromaprint":
        this.testChromaprint();
        break;
      case "dav1d":
        this.testDav1d();
        break;
      case "webp":
        this.testWebp();
        break;
      case "zscale":
        this.testZscale();
        break;
    }
  }

  testChromaprint() {
    ffprint("Testing 'chromaprint' mutex");

    this.getChromaprintSampleFile().then((audioSampleFile) {
      this.getChromaprintOutputFile().then((outputFile) {
        deleteFile(audioSampleFile);
        deleteFile(outputFile);

        final ffmpegCommand =
            "-hide_banner -y -f lavfi -i sine=frequency=1000:duration=5 -c:a pcm_s16le ${audioSampleFile.path}";

        ffprint("Creating audio sample with '${ffmpegCommand}'.");

        FFmpegKit.executeAsync(ffmpegCommand, (session) async {
          final state =
              FFmpegKitConfig.sessionStateToString(await session.getState());
          final returnCode = await session.getReturnCode();
          final failStackTrace = await session.getFailStackTrace();

          ffprint(
              "FFmpeg process exited with state ${state} and rc ${returnCode}.${notNull(failStackTrace, "\\n")}");

          if (ReturnCode.isSuccess(returnCode)) {
            ffprint("AUDIO sample created");

            final chromaprintCommand =
                "-hide_banner -y -i ${audioSampleFile.path} -f chromaprint -fp_format 2 ${outputFile.path}";

            ffprint("Creating audio sample with '${chromaprintCommand}'.");

            FFmpegKit.executeAsync(chromaprintCommand, (secondSession) async {
              final secondState = FFmpegKitConfig.sessionStateToString(
                  await secondSession.getState());
              final secondReturnCode = await secondSession.getReturnCode();
              final secondFailStackTrace =
                  await secondSession.getFailStackTrace();

              ffprint(
                  "FFmpeg process exited with state ${secondState} and rc ${secondReturnCode}.${notNull(secondFailStackTrace, "\\n")}");

              if (ReturnCode.isSuccess(secondReturnCode)) {
                showPopup("Testing chromaprint completed successfully.");
              } else {
                showPopup(
                    "Testing chromaprint failed. Please check logs for the details.");
              }
            }, (log) => this.appendOutput(log.getMessage()));
          } else {
            showPopup(
                "Creating AUDIO sample failed. Please check logs for the details.");
          }
        });
      });
    });
  }

  testDav1d() {
    ffprint("Testing decoding 'av1' codec");

    this.getDav1dOutputFile().then((outputFile) {
      final ffmpegCommand =
          "-hide_banner -y -i ${DAV1D_TEST_DEFAULT_URL} ${outputFile.path}";

      ffprint("FFmpeg process started with arguments:'${ffmpegCommand}'.");

      FFmpegKit.executeAsync(ffmpegCommand, (session) async {
        final state =
            FFmpegKitConfig.sessionStateToString(await session.getState());
        final returnCode = await session.getReturnCode();
        final failStackTrace = await session.getFailStackTrace();

        ffprint(
            "FFmpeg process exited with state ${state} and rc ${returnCode}.${notNull(failStackTrace, "\\n")}");
      }, (log) => this.appendOutput(log.getMessage()));
    });
  }

  testWebp() {
    VideoUtil.assetPath(VideoUtil.ASSET_1).then((imagePath) {
      this.getWebpOutputFile().then((outputPath) {
        ffprint("Testing 'webp' codec");

        final ffmpegCommand =
            "-hide_banner -y -i ${imagePath} ${outputPath.path}";

        ffprint("FFmpeg process started with arguments '${ffmpegCommand}'.");

        FFmpegKit.executeAsync(ffmpegCommand, (session) async {
          final state =
              FFmpegKitConfig.sessionStateToString(await session.getState());
          final returnCode = await session.getReturnCode();
          final failStackTrace = await session.getFailStackTrace();

          ffprint(
              "FFmpeg process exited with state ${state} and rc ${returnCode}.${notNull(failStackTrace, "\\n")}");

          if (ReturnCode.isSuccess(returnCode)) {
            showPopup("Encode webp completed successfully.");
          } else {
            showPopup("Encode webp failed. Please check logs for the details.");
          }
        }, (log) => this.appendOutput(log.getMessage()));
      });
    });
  }

  testZscale() {
    getVideoFile().then((videoFile) {
      getZscaledVideoFile().then((zscaledVideoFile) {
        this.getWebpOutputFile().then((outputPath) {
          ffprint(
              "Testing 'zscale' filter with video file created on the Video tab");

          final ffmpegCommand = VideoUtil.generateZscaleVideoScript(
              videoFile.path, zscaledVideoFile.path);

          ffprint("FFmpeg process started with arguments '${ffmpegCommand}'.");

          FFmpegKit.executeAsync(ffmpegCommand, (session) async {
            final state =
                FFmpegKitConfig.sessionStateToString(await session.getState());
            final returnCode = await session.getReturnCode();
            final failStackTrace = await session.getFailStackTrace();

            ffprint(
                "FFmpeg process exited with state ${state} and rc ${returnCode}.${notNull(failStackTrace, "\\n")}");

            if (ReturnCode.isSuccess(returnCode)) {
              showPopup("zscale completed successfully.");
            } else {
              showPopup("zscale failed. Please check logs for the details.");
            }
          }, (log) => this.appendOutput(log.getMessage()));
        });
      });
    });
  }

  Future<File> getChromaprintSampleFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/audio-sample.wav");
  }

  Future<File> getDav1dOutputFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/video.mp4");
  }

  Future<File> getChromaprintOutputFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/chromaprint.txt");
  }

  Future<File> getWebpOutputFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/video.webp");
  }

  Future<File> getVideoFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/video.mp4");
  }

  Future<File> getZscaledVideoFile() async {
    Directory documentsDirectory = await VideoUtil.documentsDirectory;
    return new File("${documentsDirectory.path}/video.zscaled.mp4");
  }

  List<DropdownMenuItem<String>> getTestList() {
    List<DropdownMenuItem<String>> list = List.empty(growable: true);

    list.add(new DropdownMenuItem(
        value: "chromaprint",
        child: SizedBox(
            width: 100, child: Center(child: new Text("chromaprint")))));
    list.add(new DropdownMenuItem(
        value: "dav1d",
        child: SizedBox(width: 100, child: Center(child: new Text("dav1d")))));
    list.add(new DropdownMenuItem(
        value: "webp",
        child: SizedBox(width: 100, child: Center(child: new Text("webp")))));
    list.add(new DropdownMenuItem(
        value: "zscale",
        child: SizedBox(width: 100, child: Center(child: new Text("zscale")))));

    return list;
  }

  String getOutputText() => _outputText;

  String getSelectedCodec() => _selectedTest;
}

File: linux/test-app-local-dependency/src/OtherTab.cpp
/*
 * Copyright (c) 2022 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "OtherTab.h"
#include "Application.h"
#include "Constants.h"
#include "Popup.h"
#include "Video.h"
#include <FFmpegKit.h>
#include <FFmpegKitConfig.h>

using namespace ffmpegkit;

static gboolean showTestSuccessPopup(const std::pair<Gtk::Window*,const std::string>* parameters) {
    Gtk::Window* window = parameters->first;
    auto messageDetail = parameters->second;
    ffmpegkittest::Popup::show(window, Gtk::MESSAGE_INFO, messageDetail);
    delete parameters;
    return FALSE;
}

static gboolean showTestFailedPopup(const std::pair<Gtk::Window*,const std::string>* parameters) {
    Gtk::Window* window = parameters->first;
    auto messageDetail = parameters->second;
    ffmpegkittest::Popup::show(window, Gtk::MESSAGE_ERROR, messageDetail);
    delete parameters;
    return FALSE;
}

static gboolean appendLog(const std::pair<ffmpegkittest::OtherTab*,const std::shared_ptr<Log>>* parameters) {
    ffmpegkittest::OtherTab* otherTab = parameters->first;
    auto log = parameters->second;
    otherTab->appendOutput(log->getMessage());
    delete parameters;
    return FALSE;
}

ffmpegkittest::OtherTab::OtherTab() : selectedTest(-1) {
    testModel = Gtk::ListStore::create(testModelColumn);
    test.set_model(testModel);
    test.set_size_request(240, 30);
    test.signal_changed().connect(sigc::mem_fun(*this, &OtherTab::onTestChanged));
    Util::applyComboBoxStyle(test);

    initTestData();

    runButton.set_label("RUN");
    runButton.set_size_request(120, 30);
    runButton.set_tooltip_text(Constants::OtherTestTooltipText);
    runButton.signal_clicked().connect(sigc::mem_fun(*this, &OtherTab::runTest));
    Util::applyButtonStyle(runButton);
    runButtonBox.pack_start(runButton, Gtk::PACK_EXPAND_PADDING);

    outputText.set_editable(false);
    Util::applyOutputTextStyle(outputText);
    outputTextWindow.add(outputText);

    pack_start(testBox, Gtk::PACK_SHRINK);
    pack_start(runButtonBox, Gtk::PACK_SHRINK);
    add(outputTextWindow);
}

void ffmpegkittest::OtherTab::setActive() {
    std::cout << "Other Tab Activated" << std::endl;
    FFmpegKitConfig::enableLogCallback(nullptr);
    FFmpegKitConfig::enableStatisticsCallback(nullptr);
}

void ffmpegkittest::OtherTab::setParentWindow(Gtk::Window* parentWindow) {
    this->parentWindow = parentWindow;
}

void ffmpegkittest::OtherTab::appendOutput(const std::string& string) {
    outputText.get_buffer()->set_text(outputText.get_buffer()->get_text() + string);
    Glib::RefPtr<Gtk::Adjustment> adj = outputText.get_vadjustment();
    adj->set_value(adj->get_upper());
}

void ffmpegkittest::OtherTab::clearOutput() {
    outputText.get_buffer()->set_text("");
}

void ffmpegkittest::OtherTab::initTestData() {
    auto row = *(testModel->append());
    row[testModelColumn.columnId] = "1";
    row[testModelColumn.columnName] = "chromaprint";

    row = *(testModel->append());
    row[testModelColumn.columnId] = "2";
    row[testModelColumn.columnName] = "dav1d";

    row = *(testModel->append());
    row[testModelColumn.columnId] = "3";
    row[testModelColumn.columnName] = "webp";

    row = *(testModel->append());
    row[testModelColumn.columnId] = "4";
    row[testModelColumn.columnName] = "zscale";

    test.pack_start(testModelColumn.columnName);
    test.set_entry_text_column(testModelColumn.columnId);
    test.set_active(0);

    testBox.pack_start(test, Gtk::PACK_EXPAND_PADDING);
}

void ffmpegkittest::OtherTab::onTestChanged() {
    int rowNumber = test.get_active_row_number();
    if (rowNumber != -1) {
        selectedTest = rowNumber;
    }
}

std::string ffmpegkittest::OtherTab::getSelectedTest() {
    switch(selectedTest) {
        case 0: return "chromaprint";
        case 1: return "dav1d";
        case 2: return "webp";
        case 3: return "zscale";
        default: return "";
    }
}

void ffmpegkittest::OtherTab::runTest() {
    clearOutput();

    std::string selectedTest = this->getSelectedTest();
    if (selectedTest.compare("chromaprint") == 0) {
        testChromaprint();
    } else if (selectedTest.compare("dav1d") == 0) {
        testDav1d();
    } else if (selectedTest.compare("webp") == 0) {
        testWebp();
    } else if (selectedTest.compare("zscale") == 0) {
        testZscale();
    }
}

void ffmpegkittest::OtherTab::testChromaprint() {
    std::cout << "Testing 'chromaprint' mutex." << std::endl;

    std::string audioSampleFile = getChromaprintSampleFile();
    std::remove(audioSampleFile.c_str());

    std::string ffmpegCommand = "-hide_banner -y -f lavfi -i sine=frequency=1000:duration=5 -c:a pcm_s16le " + audioSampleFile;

    std::cout << "Creating audio sample with '" << ffmpegCommand << "'." << std::endl;

    FFmpegKit::executeAsync(ffmpegCommand, [this,audioSampleFile](auto session) {
        std::cout << "FFmpeg process exited with state " << FFmpegKitConfig::sessionStateToString(session->getState()) << " and rc " << session->getReturnCode() << "." << session->getFailStackTrace() << std::endl;

        if (ReturnCode::isSuccess(session->getReturnCode())) {
            std::cout << "AUDIO sample created." << std::endl;

            std::string chromaprintCommand = "-hide_banner -y -i " + audioSampleFile + " -f chromaprint -fp_format 2 " + getChromaprintOutputFile();

            std::cout << "FFmpeg process started with arguments: '" << chromaprintCommand << "'." << std::endl;

            FFmpegKit::executeAsync(chromaprintCommand, [this](auto secondSession) {
                std::cout << "FFmpeg process exited with state " << FFmpegKitConfig::sessionStateToString(secondSession->getState()) << " and rc " << secondSession->getReturnCode() << "." << secondSession->getFailStackTrace() << std::endl;
                if (ReturnCode::isSuccess(secondSession->getReturnCode())) {
                    g_idle_add((GSourceFunc)showTestSuccessPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "Testing chromaprint completed successfully."));
                } else {
                    g_idle_add((GSourceFunc)showTestFailedPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "Testing chromaprint failed. Please check logs for the details."));
                }
            }, [this](auto log) {
                g_idle_add((GSourceFunc)appendLog, new std::pair<OtherTab*,const std::shared_ptr<Log>>(this, log));
            }, nullptr);

        } else {
            g_idle_add((GSourceFunc)showTestFailedPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "Creating AUDIO sample failed. Please check logs for the details."));
        }
    });
}

void ffmpegkittest::OtherTab::testDav1d() {
    std::cout << "Testing decoding 'av1' codec." << std::endl;

    std::string ffmpegCommand = std::string("-hide_banner -y -i ") + Dav1dTestDefaultUrl + " " + getDav1dOutputFile();

    std::cout << "FFmpeg process started with arguments: '" << ffmpegCommand << "'." << std::endl;

    FFmpegKit::executeAsync(ffmpegCommand, [this](auto session) {
        std::cout << "FFmpeg process exited with state " << FFmpegKitConfig::sessionStateToString(session->getState()) << " and rc " << session->getReturnCode() << "." << session->getFailStackTrace() << std::endl;
    }, [this](auto log) {
        g_idle_add((GSourceFunc)appendLog, new std::pair<OtherTab*,const std::shared_ptr<Log>>(this, log));
    }, nullptr);
}

void ffmpegkittest::OtherTab::testWebp() {
    std::string imageFile = Application::getApplicationInstallDirectory() + "/share/images/machupicchu.jpg";
    std::string outputFile = Application::getApplicationCacheDirectory() + "/video.webp";

    std::cout << "Testing 'webp' codec." << std::endl;

    std::string ffmpegCommand = "-hide_banner -y -i " + imageFile + " " + outputFile;

    std::cout << "FFmpeg process started with arguments: '" << ffmpegCommand << "'." << std::endl;

    auto session = FFmpegKit::executeAsync(ffmpegCommand, [this](auto session) {
        std::cout << "FFmpeg process exited with state " << FFmpegKitConfig::sessionStateToString(session->getState()) << " and rc " << session->getReturnCode() << "." << session->getFailStackTrace() << std::endl;

        if (ReturnCode::isSuccess(session->getReturnCode())) {
            g_idle_add((GSourceFunc)showTestSuccessPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "Encode webp completed successfully."));
        } else {
            g_idle_add((GSourceFunc)showTestFailedPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "Encode webp failed. Please check logs for the details."));
        }
    }, [this](auto log) {
        g_idle_add((GSourceFunc)appendLog, new std::pair<OtherTab*,const std::shared_ptr<Log>>(this, log));
    }, nullptr);
}

void ffmpegkittest::OtherTab::testZscale() {
    std::string videoFile = Application::getApplicationCacheDirectory() + "/video.mp4";
    std::string zscaledVideoFile = Application::getApplicationCacheDirectory() + "/video-zscaled.mp4";

    std::cout << "Testing 'zscale' filter with video file created on the Video tab." << std::endl;

    std::string ffmpegCommand = Video::generateZscaleVideoScript(videoFile, zscaledVideoFile);

    std::cout << "FFmpeg process started with arguments: '" << ffmpegCommand << "'." << std::endl;

    auto session = FFmpegKit::executeAsync(ffmpegCommand, [this](auto session) {
        std::cout << "FFmpeg process exited with state " << FFmpegKitConfig::sessionStateToString(session->getState()) << " and rc " << session->getReturnCode() << "." << session->getFailStackTrace() << std::endl;

        if (ReturnCode::isSuccess(session->getReturnCode())) {
            g_idle_add((GSourceFunc)showTestSuccessPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "zscale completed successfully."));
        } else {
            g_idle_add((GSourceFunc)showTestFailedPopup, new std::pair<Gtk::Window*,const std::string>(this->parentWindow, "zscale failed. Please check logs for the details."));
        }
    }, [this](auto log) {
        g_idle_add((GSourceFunc)appendLog, new std::pair<OtherTab*,const std::shared_ptr<Log>>(this, log));
    }, nullptr);
}

std::string ffmpegkittest::OtherTab::getChromaprintSampleFile() {
    return Application::getApplicationCacheDirectory() + "/audio-sample.wav";
}

std::string ffmpegkittest::OtherTab::getDav1dOutputFile() {
    return Application::getApplicationCacheDirectory() + "/video.mp4";
}

std::string ffmpegkittest::OtherTab::getChromaprintOutputFile() {
    return Application::getApplicationCacheDirectory() + "/chromaprint.txt";
}
